blockcache/blockcache.go

```diff
@@ -0,0 +1,224 @@
+package blockcache
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/boltdb/bolt"
+	"github.com/lightninglabs/neutrino/headerfs"
+	"github.com/roasbeef/btcd/chaincfg/chainhash"
+	"github.com/roasbeef/btcd/wire"
+	"os"
+	"path/filepath"
+	"strconv"
+	"sync/atomic"
+)
+
+var (
+	// ErrBlockNotFound is returned when the requested block is not persisted.
+	ErrBlockNotFound = fmt.Errorf("unable to find block")
+
+	// blockCacheBucket is the name of the database bucket we'll use to
+	// persist the blocks.
+	blockCacheBucket = []byte("block-cache")
+
+	// DefaultCapacity is the default maximum capacity of the
+	// MostRecentBlockCache, in blocks.
+	DefaultCapacity = 1000
+)
+
+const (
+	dbName           = "blockcache.db"
+	dbFilePermission = 0600
+)
+
+// BlockCache is an interface which represents an object that is capable of
+// storing and retrieving blocks according to their corresponding block hash.
+//
+type BlockCache interface {
+	// PutBlock stores a block to persistent storage.
+	PutBlock(block *wire.MsgBlock) error
+
+	// FetchBlock attempts to fetch a block with the given hash from the
+	// backing store. If the backing store does not contain the block then
+	// an ErrBlockNotFound error is returned.
+	FetchBlock(*chainhash.Hash) (*wire.MsgBlock, error)
+
+	// Close releases any resources currently in use by the cache.
+	Close()
+}
+
+// MostRecentBlockCache is an implementation of the BlockCache interface
+// which attempts to cache a fixed number of the most recent blocks.
+// If a newer block (by height) is inserted and the cache is already full,
+// then the oldest block will be evicted to make room.
+type MostRecentBlockCache struct {
+	db *bolt.DB
+
+	// A BlockHeaderStore allows translation between block heights and hashes.
+	blockHeaders *headerfs.BlockHeaderStore
+
+	// The maximum number of blocks to store in the cache.
+	capacity int
+
+	// The number of cache hits and misses. Must be accessed atomically.
+	hits   uint64
```

Comment:
These should be moved to the top of the struct in order to avoid issues accessing them atomically 32-bit platforms.